ios/LayoutReanimation/REASharedTransitionManager.m

```diff
@@ -0,0 +1,518 @@
+                                                                           withNewElements:NO];
+  if ([sharedElements count] == 0) {
+    return NO;
+  }
+  [self setupTransitionContainer];
+  [self reparentSharedViewsForCurrentTransition:sharedElements];
+  [self startSharedTransition:sharedElements];
+  return YES;
+}
+
+- (NSArray *)sortViewsByTags:(NSArray *)views
+{
+  /*
+    All shared views during the transition have the same parent. It is problematic if parent
+    view and their children are in the same transition. To keep the valid order in the z-axis,
+    we need to sort views by tags. Parent tag is lower than children tags.
+  */
+  return [views sortedArrayUsingComparator:^NSComparisonResult(UIView *view1, UIView *view2) {
+    return [view2.reactTag compare:view1.reactTag];
+  }];
+}
+
+- (NSMutableArray<REASharedElement *> *)getSharedElementForCurrentTransition:(NSArray *)sharedViews
+                                                             withNewElements:(BOOL)addedNewScreen
+{
+  NSMutableArray<REASharedElement *> *sharedElements = [NSMutableArray new];
+  for (UIView *sharedView in sharedViews) {
+    // add observers
+    UIView *sharedViewScreen = [self getScreenForView:sharedView];
+    UIView *stack = [self getStackForView:sharedViewScreen];
+    if (addedNewScreen) {
+      [self observeChanges:sharedViewScreen];
+    }
+
+    // find sibling for shared view
+    NSNumber *siblingViewTag = _findPrecedingViewTagForTransition(sharedView.reactTag);
+    UIView *siblingView = nil;
+    do {
+      siblingView = [_animationManager viewForTag:siblingViewTag];
+      if (siblingView == nil) {
+        [self clearAllSharedConfigsForViewTag:siblingViewTag];
+        siblingViewTag = _findPrecedingViewTagForTransition(sharedView.reactTag);
+      }
+    } while (siblingView == nil && siblingViewTag != nil);
+
+    if (siblingView == nil) {
+      // the sibling of shared view doesn't exist yet
+      continue;
+    }
+
+    UIView *viewSource;
+    UIView *viewTarget;
+    if (addedNewScreen) {
+      viewSource = siblingView;
+      viewTarget = sharedView;
+    } else {
+      viewSource = sharedView;
+      viewTarget = siblingView;
+    }
+
+    // check valid target screen configuration
+    int screensCount = [stack.reactSubviews count];
+    if (addedNewScreen) {
+      // is under top
+      if (screensCount < 2) {
+        continue;
+      }
+      UIView *viewSourceParentScreen = [self getScreenForView:viewSource];
+      UIView *screenUnderStackTop = stack.reactSubviews[screensCount - 2];
+      if (![screenUnderStackTop.reactTag isEqual:viewSourceParentScreen.reactTag]) {
+        continue;
+      }
+    } else {
+      // is on top
+      UIView *viewTargetParentScreen = [self getScreenForView:viewTarget];
+      UIView *stackTarget = viewTargetParentScreen.reactViewController.navigationController.topViewController.view;
+      if (stackTarget != viewTargetParentScreen) {
+        continue;
+      }
+    }
```

Comment:
I wonder - should this be inside the `for`-loop over transitioning views? Can we start a transition between different pairs of screens at the same time, or should they always be a transition of views on screen A to views on screen B, in which case this could be done once, outside the loop?

Reply:
In this place, I check if the animation should be run. The `_findPrecedingViewTagForTransition` return just the view tag register before the current shared view, but the pair of views is doesn't always valid. For example for screen stack:
2: ScreenC - has a shared view 
1: ScreenB - hasn't a shared view
0: ScreenA - has a shared view

And if we want to go back from ScreenC -> ScreenB, `_findPrecedingViewTagForTransition` returns a view tag from ScreenA but we don't want to run animation in this case.